### PR TITLE
Misc

### PR DESCRIPTION
**Pull request by regislebrun**
This PR contains the following features:
  * a slight improvement in the sampling of Normal and ComposedDistribution, resulting into a gain of 5% in the sampling speed
  * a fix of the Normal class, see https://github.com/openturns/openturns/issues/2463
  * a fix for Triangular and Trapezoidal characteristic functions with small argument, see https://github.com/openturns/openturns/issues/2216

### CODE REVIEW

File: lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx [hunk 1]
```diff
@@ -302,14 +302,39 @@ void EfficientGlobalOptimization::run()
...
+    if (!problem.hasBounds() || problem.getBounds().contains(newPoint))
+      {
+        improvementValue = improvementResult.getOptimalValue();
+        expectedImprovement_.add(improvementValue);
+
+        newValue = model(newPoint);
+        ++ evaluationNumber;
+      }
+    else
+    // Restore the admissibility wrt bounds if needed
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx:318]: This should not be needed if enforced in other algos, see if my last commit in misc2 does what you want

File: lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx [hunk 2]
```diff
@@ -302,14 +302,39 @@ void EfficientGlobalOptimization::run()
...
-    Point improvementValue(improvementResult.getOptimalValue());
-    expectedImprovement_.add(improvementValue);
+    Point newPoint(improvementResult.getOptimalPoint());
+    Point newValue;
+    Point improvementValue;
+    if (!problem.hasBounds() || problem.getBounds().contains(newPoint))
+      {
+        improvementValue = improvementResult.getOptimalValue();
+        expectedImprovement_.add(improvementValue);
+
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx:313]:
```suggestion
```

File: lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx [hunk 3]
```diff
@@ -302,14 +302,39 @@ void EfficientGlobalOptimization::run()
...
     // store improvement
-    Point improvementValue(improvementResult.getOptimalValue());
-    expectedImprovement_.add(improvementValue);
+    Point newPoint(improvementResult.getOptimalPoint());
+    Point newValue;
+    Point improvementValue;
+    if (!problem.hasBounds() || problem.getBounds().contains(newPoint))
+      {
+        improvementValue = improvementResult.getOptimalValue();
+        expectedImprovement_.add(improvementValue);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx:312]: This is not the topic of your development, but the naming is as confusing as can be. This is not `expectedImprovement_` but `expectedImprovementHistory_`. And the method `EfficientGlobalOptimization::getExpectedImprovement` should be renamed `EfficientGlobalOptimization::getExpectedImprovementHistory`.

File: lib/src/Uncertainty/Distribution/Trapezoidal.cxx [hunk 4]
```diff
@@ -168,8 +168,8 @@ Complex Trapezoidal::computeCharacteristicFunction(const Scalar u) const
   Scalar real = 0.0;
   Scalar imag = 0.0;
   Complex phi;
-
-  if (std::abs(u) <= 1.0e-10) phi = Complex((-a_ + 3.0 * b_ - 3.0 * c_ + d_) * h_ / 2.0, 1.0);
+  const Scalar factor = h_ * (((c_ + d_) * (c_ * c_ * c_ + d_ * d_ * d_) + c_ * c_ * d_ * d_) - ((a_ + b_) * (a_ * a_ * a_ + b_ * b_ * b_) + a_ * a_ * b_ * b_)) / 120.0;
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/Trapezoidal.cxx:171]: We may inadvertently break that code without noticing the change. This cannot happen if you provide a unit test that shows the improvement in action.

File: lib/src/Uncertainty/Distribution/Triangular.cxx [hunk 5]
```diff
@@ -162,8 +162,9 @@ Scalar Triangular::computeCDF(const Point & point) const
 /* Get the characteristic function of the distribution, i.e. phi(u) = E(exp(I*u*X)) */
 Complex Triangular::computeCharacteristicFunction(const Scalar x) const
 {
-  if (std::abs(x) < pdfEpsilon_) return 1.0;
-  if (std::abs(x) < 1.0e-8) return Complex(1.0, (a_ + b_ + m_) * x / 3.0);
+  // Here we need a high order approximation as the exact formula is ill conditioned for small values of x
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/Triangular.cxx:165]: A unit test will make this update more robust.

**Review by mbaudin47 — CHANGES_REQUESTED** [review]
I think that 2 new unit tests will improve the Triangular part of the PR.

File: lib/src/Uncertainty/Distribution/Mixture.cxx [hunk 6]
```diff
@@ -333,7 +345,7 @@ Scalar Mixture::computeSurvivalFunction(const Point & point) const
   Scalar survivalValue = 0.0;
   const UnsignedInteger size = distributionCollection_.getSize();
   for(UnsignedInteger i = 0; i < size; ++i) survivalValue += p_[i] * distributionCollection_[i].computeSurvivalFunction(point);
-  return survivalValue;
+  return std::min(1.0, std::max(0.0, survivalValue));
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Mixture.cxx:348]: you can use SpecFunc::Clip01

File: lib/src/Uncertainty/Distribution/Mixture.cxx [hunk 7]
```diff
@@ -321,7 +321,19 @@ Scalar Mixture::computeCDF(const Point & point) const
...
+/* Get the complementary CDF of the Mixture */
+Scalar Mixture::computeComplementaryCDF(const Point & point) const
+{
+  const UnsignedInteger dimension = getDimension();
+  if (point.getDimension() != dimension) throw InvalidArgumentException(HERE) << "Error: the given point must have dimension=" << dimension << ", here dimension=" << point.getDimension();
+
+  Scalar complementaryCDFValue = 0.0;
+  const UnsignedInteger size = distributionCollection_.getSize();
+  for(UnsignedInteger i = 0; i < size; ++i) complementaryCDFValue += p_[i] * distributionCollection_[i].computeComplementaryCDF(point);
+  return std::min(1.0, std::max(0.0, complementaryCDFValue));
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Mixture.cxx:336]: here too

File: lib/src/Uncertainty/Distribution/Mixture.cxx [hunk 8]
```diff
@@ -321,7 +321,19 @@ Scalar Mixture::computeCDF(const Point & point) const
   Scalar cdfValue = 0.0;
   const UnsignedInteger size = distributionCollection_.getSize();
   for(UnsignedInteger i = 0; i < size; ++i) cdfValue += p_[i] * distributionCollection_[i].computeCDF(point);
-  return cdfValue;
+  return std::min(1.0, std::max(0.0, cdfValue));
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Mixture.cxx:324]: and here

File: lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx (hunk 2 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx]: The commit has been removed from the PR, the issue being solved in another PR.

File: lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx (hunk 3 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx]: The commit has been removed from the PR, the issue being solved in another PR.

File: lib/src/Uncertainty/Distribution/Trapezoidal.cxx (hunk 4 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Trapezoidal.cxx]: -Excellent suggestion. Done in t_Triangular_std.cxx and t_Trapezoidal_std.cxx, using Maple reference values (see validation/srd/valid_triangular_cf.mws and validation/srd/valid_trapezoidal_cf.mws)

File: lib/src/Uncertainty/Distribution/Triangular.cxx (hunk 5 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Triangular.cxx:165]: Excellent suggestion. Done in t_Triangular_std.cxx and t_Trapezoidal_std.cxx, using Maple reference values (see validation/srd/valid_triangular_cf.mws and validation/srd/valid_trapezoidal_cf.mws)

File: lib/src/Uncertainty/Distribution/Mixture.cxx (hunk 6 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Mixture.cxx]: Done

File: lib/src/Uncertainty/Distribution/Mixture.cxx (hunk 7 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Mixture.cxx]: Done

File: lib/src/Uncertainty/Distribution/Mixture.cxx (hunk 8 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Mixture.cxx]: Done

File: lib/test/t_Triangular_std.cxx [hunk 9]
```diff
@@ -104,6 +127,8 @@ int main(int, char *[])
       fullprint << "characteristic function=" << CF << std::endl;
       Complex LCF = distribution.computeLogCharacteristicFunction( point[0] );
       fullprint << "log characteristic function=" << LCF << std::endl;
+      for (UnsignedInteger j = 0; j < refValues[nTriangular].getSize(); ++j)
+        assert(distribution.computeCharacteristicFunction(u[j], refValues[nTriangular][j]));
```
Comment (jschueller) [inline lib/test/t_Triangular_std.cxx:131]: this doesnt compile
did you mean assert_almost_equal ?
the parenthesis of computeCharacteristicFunction is misplaced
same in Trapezoidal test

Comment (regislebrun) [reply lib/test/t_Triangular_std.cxx]: I saw it. I pushed an unsaved file :-(
It should be fixed now...

**Review by jschueller — APPROVED** [review]

**Pull request merged**